src/twisted/trial/_asynctest.py

```diff
@@ -134,43 +138,43 @@ def onTimeout(d):
     def __call__(self, *args, **kwargs):
         return self.run(*args, **kwargs)
 
-    def _deferSetUp(self, ignored, result):
-        d = self._run(self.setUp, "setUp", result)
-        d.addCallbacks(
-            self._deferTestMethod,
-            self._ebDeferSetUp,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        return d
+    async def _deferSetUp(self, result):
+        try:
+            try:
+                await self._deferSetUpNoCleanup(result)
+            finally:
+                await self._deferRunCleanups(result)
+        finally:
+            if self._twistedPrivateNeedsTearDown:
+                await self._deferTearDown(result)
 
-    def _ebDeferSetUp(self, failure, result):
-        if failure.check(SkipTest):
-            result.addSkip(self, self._getSkipReason(self.setUp, failure.value))
-        else:
-            result.addError(self, failure)
-            if failure.check(KeyboardInterrupt):
-                result.stop()
-        return self._deferRunCleanups(None, result)
-
-    def _deferTestMethod(self, ignored, result):
-        d = self._run(getattr(self, self._testMethodName), self._testMethodName, result)
-        d.addCallbacks(
-            self._cbDeferTestMethod,
-            self._ebDeferTestMethod,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        d.addBoth(self._deferRunCleanups, result)
-        d.addBoth(self._deferTearDown, result)
-        return d
+    async def _deferSetUpNoCleanup(self, result):
```

Comment:
Maybe this is a better name


```suggestion
    async def _deferSetUpAndRun(self, result):
        """
        Execute the setUp and run part of a test.

        Teardown and cleanups are not executed.
        """
```

Reply:
Applied now, thanks.